In [1]:
# error=> module not found represents a library error

In [1]:
!pip install pandas

In [2]:
import pandas as pd

In [3]:
dir(pd)

['ArrowDtype',
 'BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt8Dtype',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_built_with_meson',
 '_config',
 '_is_numpy_dev',
 '_libs',
 '_pandas_datetime_CAPI',
 '_pandas_parser_CAPI',
 '_testing',
 '_typing',
 '_version_meson',
 'annota

In [4]:
df = pd.read_csv('retail supermarket.csv')
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

To find rows & columns in the file.

In [5]:
df.shape

(9994, 13)

In [2]:
# shape is an attribute bcz there is no parenthesis, doesn't take any argument.

0 in shape represents number of rows whereas 1 respresents number of columns.

In [6]:
df.shape[0]

9994

In [7]:
df.shape[1]

13

In [4]:
# info provide us all the relevent info of the data.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1015.1+ KB


In [9]:
# Pandas doesn't have string in its datatype, it obtains object inreplacement of string.

In [3]:
# dtypes tell us about the types of each column.

In [10]:
df.dtypes

Ship Mode        object
Segment          object
Country          object
City             object
State            object
Postal Code       int64
Region           object
Category         object
Sub-Category     object
Sales           float64
Quantity          int64
Discount        float64
Profit          float64
dtype: object

head respresents the above 5 rows of data, bydefault.

In [11]:
df.head()

Ship Mode    Segment        Country             City       State  \
0    Second Class   Consumer  United States        Henderson    Kentucky   
1    Second Class   Consumer  United States        Henderson    Kentucky   
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
0        42420  South        Furniture    Bookcases  261.9600         2   
1        42420  South        Furniture       Chairs  731.9400         3   
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164

In [12]:
df.head(10)

Ship Mode    Segment        Country             City       State  \
0    Second Class   Consumer  United States        Henderson    Kentucky   
1    Second Class   Consumer  United States        Henderson    Kentucky   
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
0        42420  South        Furniture    Bookcases  261.9600         2   
1        42420  South        Furniture       Chairs  731.9400         3   
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   
5        90032   West        Furniture  Furnishings   48.8600         7   
6        90032   West  Office Supplies          Art    7.2800         4   
7        90032   West       Technology       Phones  907.1520         6   
8        90032   West  Office Supplies      Binders   18.5040         3   
9        90032   West  Office Supplies   Appliances  114.9000         5   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700

tail respresents the data on ending 5 rows, bydefault.

In [13]:
df.tail()

Ship Mode   Segment        Country         City       State  \
9989    Second Class  Consumer  United States        Miami     Florida   
9990  Standard Class  Consumer  United States   Costa Mesa  California   
9991  Standard Class  Consumer  United States   Costa Mesa  California   
9992  Standard Class  Consumer  United States   Costa Mesa  California   
9993    Second Class  Consumer  United States  Westminster  California   

      Postal Code Region         Category Sub-Category    Sales  Quantity  \
9989        33180  South        Furniture  Furnishings   25.248         3   
9990        92627   West        Furniture  Furnishings   91.960         2   
9991        92627   West       Technology       Phones  258.576         2   
9992        92627   West  Office Supplies        Paper   29.600         4   
9993        92683   West  Office Supplies   Appliances  243.160         2   

      Discount   Profit  
9989       0.2   4.1028  
9990       0.0  15.6332  
9991       0.2  19.3932  
9992       0.0  13.3200  
9993       0.0  72.9480

In [14]:
df.tail(15)

Ship Mode      Segment        Country          City       State  \
9979  Standard Class  Home Office  United States   Los Angeles  California   
9980    Second Class     Consumer  United States     Lafayette   Louisiana   
9981     First Class     Consumer  United States     Fairfield        Ohio   
9982  Standard Class     Consumer  United States  Grand Rapids    Michigan   
9983  Standard Class     Consumer  United States  Grand Rapids    Michigan   
9984  Standard Class     Consumer  United States    Long Beach    New York   
9985  Standard Class     Consumer  United States    Long Beach    New York   
9986  Standard Class     Consumer  United States   Los Angeles  California   
9987  Standard Class    Corporate  United States        Athens     Georgia   
9988  Standard Class    Corporate  United States        Athens     Georgia   
9989    Second Class     Consumer  United States         Miami     Florida   
9990  Standard Class     Consumer  United States    Costa Mesa  California   
9991  Standard Class     Consumer  United States    Costa Mesa  California   
9992  Standard Class     Consumer  United States    Costa Mesa  California   
9993    Second Class     Consumer  United States   Westminster  California   

      Postal Code   Region         Category Sub-Category    Sales  Quantity  \
9979        90032     West  Office Supplies      Binders  437.472        14   
9980        70506    South        Furniture       Tables   85.980         1   
9981        45014     East  Office Supplies       Labels   16.520         5   
9982        49505  Central  Office Supplies        Paper   35.560         7   
9983        49505  Central       Technology       Phones   97.980         2   
9984        11561     East  Office Supplies       Labels   31.500        10   
9985        11561     East  Office Supplies     Supplies   55.600         4   
9986        90008     West       Technology  Accessories   36.240         1   
9987        30605    South       Technology  Accessories   79.990         1   
9988        30605    South       Technology       Phones  206.100         5   
9989        33180    South        Furniture  Furnishings   25.248         3   
9990        92627     West        Furniture  Furnishings   91.960         2   
9991        92627     West       Technology       Phones  258.576         2   
9992        92627     West  Office Supplies        Paper   29.600         4   
9993        92683     West  Office Supplies   Appliances  243.160         2   

      Discount    Profit  
9979       0.2  153.1152  
9980       0.0   22.3548  
9981       0.2    5.3690  
9982       0.0   16.7132  
9983       0.0   27.4344  
9984       0.0   15.1200  
9985       0.0   16.1240  
9986       0.0   15.2208  
9987       0.0   28.7964  
9988       0.0   55.6470  
9989       0.2    4.1028  
9990       0.0   15.6332  
9991       0.2   19.3932  
9992       0.0   13.3200  
9993       0.0   72.9480

In Pandas we do indexing through loc & iloc.

**loc (Label-based indexing)** 
Syntax: `df.loc[row_label, column_label]`

In [15]:
# syntax, the way to write the program. 

In [16]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

In [17]:
df.loc[2]

Ship Mode          Second Class
Segment               Corporate
Country           United States
City                Los Angeles
State                California
Postal Code               90036
Region                     West
Category        Office Supplies
Sub-Category             Labels
Sales                     14.62
Quantity                      2
Discount                    0.0
Profit                   6.8714
Name: 2, dtype: object

In [18]:
df.loc[2,'Postal Code']

90036

**iloc (Integer-based indexing)** `Syntax: df.iloc[row_index, column_index]`

In [19]:
df.iloc[2]

Ship Mode          Second Class
Segment               Corporate
Country           United States
City                Los Angeles
State                California
Postal Code               90036
Region                     West
Category        Office Supplies
Sub-Category             Labels
Sales                     14.62
Quantity                      2
Discount                    0.0
Profit                   6.8714
Name: 2, dtype: object

In [20]:
df.iloc[2,5]

90036

In [21]:
df.loc[2,5]

KeyError: 5

In [22]:
# bcz you're writing loc and giving instructions of columnn in integer as you're working in iloc.

In [23]:
df.loc[2:10]

Ship Mode    Segment        Country             City       State  \
2     Second Class  Corporate  United States      Los Angeles  California   
3   Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4   Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5   Standard Class   Consumer  United States      Los Angeles  California   
6   Standard Class   Consumer  United States      Los Angeles  California   
7   Standard Class   Consumer  United States      Los Angeles  California   
8   Standard Class   Consumer  United States      Los Angeles  California   
9   Standard Class   Consumer  United States      Los Angeles  California   
10  Standard Class   Consumer  United States      Los Angeles  California   

    Postal Code Region         Category Sub-Category      Sales  Quantity  \
2         90036   West  Office Supplies       Labels    14.6200         2   
3         33311  South        Furniture       Tables   957.5775         5   
4         33311  South  Office Supplies      Storage    22.3680         2   
5         90032   West        Furniture  Furnishings    48.8600         7   
6         90032   West  Office Supplies          Art     7.2800         4   
7         90032   West       Technology       Phones   907.1520         6   
8         90032   West  Office Supplies      Binders    18.5040         3   
9         90032   West  Office Supplies   Appliances   114.9000         5   
10        90032   West        Furniture       Tables  1706.1840         9   

    Discount    Profit  
2       0.00    6.8714  
3       0.45 -383.0310  
4       0.20    2.5164  
5       0.00   14.1694  
6       0.00    1.9656  
7       0.20   90.7152  
8       0.20    5.7825  
9       0.00   34.4700  
10      0.20   85.3092

In [24]:
# label indexing provides output till you mention it.

In [25]:
df.iloc[2:10]

Ship Mode    Segment        Country             City       State  \
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   
5        90032   West        Furniture  Furnishings   48.8600         7   
6        90032   West  Office Supplies          Art    7.2800         4   
7        90032   West       Technology       Phones  907.1520         6   
8        90032   West  Office Supplies      Binders   18.5040         3   
9        90032   West  Office Supplies   Appliances  114.9000         5   

   Discount    Profit  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700

In [26]:
# integer indexing provides output till before the last one you mention.

In [27]:
df.iloc[2:10,:7]

Ship Mode    Segment        Country             City       State  \
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region  
2        90036   West  
3        33311  South  
4        33311  South  
5        90032   West  
6        90032   West  
7        90032   West  
8        90032   West  
9        90032   West

In [28]:
df.loc[2:9]

Ship Mode    Segment        Country             City       State  \
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   
5        90032   West        Furniture  Furnishings   48.8600         7   
6        90032   West  Office Supplies          Art    7.2800         4   
7        90032   West       Technology       Phones  907.1520         6   
8        90032   West  Office Supplies      Binders   18.5040         3   
9        90032   West  Office Supplies   Appliances  114.9000         5   

   Discount    Profit  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700

In [29]:
df.loc[2:9,:'Region']

Ship Mode    Segment        Country             City       State  \
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region  
2        90036   West  
3        33311  South  
4        33311  South  
5        90032   West  
6        90032   West  
7        90032   West  
8        90032   West  
9        90032   West

To make segment as index of the data.

In [30]:
df1 = pd.read_csv('retail supermarket.csv',index_col=1)
df1

Ship Mode        Country             City       State  \
Segment                                                                 
Consumer     Second Class  United States        Henderson    Kentucky   
Consumer     Second Class  United States        Henderson    Kentucky   
Corporate    Second Class  United States      Los Angeles  California   
Consumer   Standard Class  United States  Fort Lauderdale     Florida   
Consumer   Standard Class  United States  Fort Lauderdale     Florida   
...                   ...            ...              ...         ...   
Consumer     Second Class  United States            Miami     Florida   
Consumer   Standard Class  United States       Costa Mesa  California   
Consumer   Standard Class  United States       Costa Mesa  California   
Consumer   Standard Class  United States       Costa Mesa  California   
Consumer     Second Class  United States      Westminster  California   

           Postal Code Region         Category Sub-Category     Sales  \
Segment                                                                 
Consumer         42420  South        Furniture    Bookcases  261.9600   
Consumer         42420  South        Furniture       Chairs  731.9400   
Corporate        90036   West  Office Supplies       Labels   14.6200   
Consumer         33311  South        Furniture       Tables  957.5775   
Consumer         33311  South  Office Supplies      Storage   22.3680   
...                ...    ...              ...          ...       ...   
Consumer         33180  South        Furniture  Furnishings   25.2480   
Consumer         92627   West        Furniture  Furnishings   91.9600   
Consumer         92627   West       Technology       Phones  258.5760   
Consumer         92627   West  Office Supplies        Paper   29.6000   
Consumer         92683   West  Office Supplies   Appliances  243.1600   

           Quantity  Discount    Profit  
Segment                                  
Consumer          2      0.00   41.9136  
Consumer          3      0.00  219.5820  
Corporate         2      0.00    6.8714  
Consumer          5      0.45 -383.0310  
Consumer          2      0.20    2.5164  
...             ...       ...       ...  
Consumer          3      0.20    4.1028  
Consumer          2      0.00   15.6332  
Consumer          2      0.20   19.3932  
Consumer          4      0.00   13.3200  
Consumer          2      0.00   72.9480  

[9994 rows x 12 columns]

In [31]:
df1.loc['Consumer']

Ship Mode        Country             City       State  \
Segment                                                                
Consumer    Second Class  United States        Henderson    Kentucky   
Consumer    Second Class  United States        Henderson    Kentucky   
Consumer  Standard Class  United States  Fort Lauderdale     Florida   
Consumer  Standard Class  United States  Fort Lauderdale     Florida   
Consumer  Standard Class  United States      Los Angeles  California   
...                  ...            ...              ...         ...   
Consumer    Second Class  United States            Miami     Florida   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer    Second Class  United States      Westminster  California   

          Postal Code Region         Category Sub-Category     Sales  \
Segment                                                                
Consumer        42420  South        Furniture    Bookcases  261.9600   
Consumer        42420  South        Furniture       Chairs  731.9400   
Consumer        33311  South        Furniture       Tables  957.5775   
Consumer        33311  South  Office Supplies      Storage   22.3680   
Consumer        90032   West        Furniture  Furnishings   48.8600   
...               ...    ...              ...          ...       ...   
Consumer        33180  South        Furniture  Furnishings   25.2480   
Consumer        92627   West        Furniture  Furnishings   91.9600   
Consumer        92627   West       Technology       Phones  258.5760   
Consumer        92627   West  Office Supplies        Paper   29.6000   
Consumer        92683   West  Office Supplies   Appliances  243.1600   

          Quantity  Discount    Profit  
Segment                                 
Consumer         2      0.00   41.9136  
Consumer         3      0.00  219.5820  
Consumer         5      0.45 -383.0310  
Consumer         2      0.20    2.5164  
Consumer         7      0.00   14.1694  
...            ...       ...       ...  
Consumer         3      0.20    4.1028  
Consumer         2      0.00   15.6332  
Consumer         2      0.20   19.3932  
Consumer         4      0.00   13.3200  
Consumer         2      0.00   72.9480  

[5191 rows x 12 columns]

In [32]:
df1.loc['Consumer',:'Quantity']

Ship Mode        Country             City       State  \
Segment                                                                
Consumer    Second Class  United States        Henderson    Kentucky   
Consumer    Second Class  United States        Henderson    Kentucky   
Consumer  Standard Class  United States  Fort Lauderdale     Florida   
Consumer  Standard Class  United States  Fort Lauderdale     Florida   
Consumer  Standard Class  United States      Los Angeles  California   
...                  ...            ...              ...         ...   
Consumer    Second Class  United States            Miami     Florida   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer  Standard Class  United States       Costa Mesa  California   
Consumer    Second Class  United States      Westminster  California   

          Postal Code Region         Category Sub-Category     Sales  Quantity  
Segment                                                                         
Consumer        42420  South        Furniture    Bookcases  261.9600         2  
Consumer        42420  South        Furniture       Chairs  731.9400         3  
Consumer        33311  South        Furniture       Tables  957.5775         5  
Consumer        33311  South  Office Supplies      Storage   22.3680         2  
Consumer        90032   West        Furniture  Furnishings   48.8600         7  
...               ...    ...              ...          ...       ...       ...  
Consumer        33180  South        Furniture  Furnishings   25.2480         3  
Consumer        92627   West        Furniture  Furnishings   91.9600         2  
Consumer        92627   West       Technology       Phones  258.5760         2  
Consumer        92627   West  Office Supplies        Paper   29.6000         4  
Consumer        92683   West  Office Supplies   Appliances  243.1600         2  

[5191 rows x 10 columns]

In [33]:
# If you want to pick different rows then do it in a separate list bracket.

In [34]:
df.iloc[[2,10,3,100]]

Ship Mode      Segment        Country             City       State  \
2      Second Class    Corporate  United States      Los Angeles  California   
10   Standard Class     Consumer  United States      Los Angeles  California   
3    Standard Class     Consumer  United States  Fort Lauderdale     Florida   
100  Standard Class  Home Office  United States          Chicago    Illinois   

     Postal Code   Region         Category Sub-Category      Sales  Quantity  \
2          90036     West  Office Supplies       Labels    14.6200         2   
10         90032     West        Furniture       Tables  1706.1840         9   
3          33311    South        Furniture       Tables   957.5775         5   
100        60610  Central       Technology  Accessories    95.9760         3   

     Discount    Profit  
2        0.00    6.8714  
10       0.20   85.3092  
3        0.45 -383.0310  
100      0.20  -10.7973

In [35]:
# If you want to pick different rows and columns then do it in a separate list bracket.

In [36]:
df.loc[[2,10],['Segment','Category']]

Segment         Category
2   Corporate  Office Supplies
10   Consumer        Furniture

In [37]:
df.iloc[[2,10],[1,7]]

Segment         Category
2   Corporate  Office Supplies
10   Consumer        Furniture

In [38]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

In [39]:
df['City']

0             Henderson
1             Henderson
2           Los Angeles
3       Fort Lauderdale
4       Fort Lauderdale
             ...       
9989              Miami
9990         Costa Mesa
9991         Costa Mesa
9992         Costa Mesa
9993        Westminster
Name: City, Length: 9994, dtype: object

In [40]:
# In Pandas column is known as series whereas all columns collectively are called dataframe.

Here above can see lipses occur btw 4 & 9989 data, so to watch data as a whole type the below program.

In [41]:
#pd.options.display.max_rows = None

In [42]:
# None is for all, if you mention anything it will provide the output as per your desire.

In [43]:
# commenting the above program bcz it will show me all the data, afterwards.

In [44]:
df.head(1)

Ship Mode   Segment        Country       City     State  Postal Code  \
0  Second Class  Consumer  United States  Henderson  Kentucky        42420   

  Region   Category Sub-Category   Sales  Quantity  Discount   Profit  
0  South  Furniture    Bookcases  261.96         2       0.0  41.9136

In [45]:
df[['Country','Postal Code','Category']]

Country  Postal Code         Category
0     United States        42420        Furniture
1     United States        42420        Furniture
2     United States        90036  Office Supplies
3     United States        33311        Furniture
4     United States        33311  Office Supplies
...             ...          ...              ...
9989  United States        33180        Furniture
9990  United States        92627        Furniture
9991  United States        92627       Technology
9992  United States        92627  Office Supplies
9993  United States        92683  Office Supplies

[9994 rows x 3 columns]

In [46]:
df[['Country','Postal Code','Category']].head()

Country  Postal Code         Category
0  United States        42420        Furniture
1  United States        42420        Furniture
2  United States        90036  Office Supplies
3  United States        33311        Furniture
4  United States        33311  Office Supplies

In [48]:
# drop is to remove column.

In [47]:
df.drop('Postal Code')

KeyError: "['Postal Code'] not found in axis"

In [49]:
# here you want column to drop, so give its axis 1.

In [50]:
df.drop('Postal Code',axis = 1)

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
0     South        Furniture    Bookcases  261.9600         2      0.00   
1     South        Furniture       Chairs  731.9400         3      0.00   
2      West  Office Supplies       Labels   14.6200         2      0.00   
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
...     ...              ...          ...       ...       ...       ...   
9989  South        Furniture  Furnishings   25.2480         3      0.20   
9990   West        Furniture  Furnishings   91.9600         2      0.00   
9991   West       Technology       Phones  258.5760         2      0.20   
9992   West  Office Supplies        Paper   29.6000         4      0.00   
9993   West  Office Supplies   Appliances  243.1600         2      0.00   

        Profit  
0      41.9136  
1     219.5820  
2       6.8714  
3    -383.0310  
4       2.5164  
...        ...  
9989    4.1028  
9990   15.6332  
9991   19.3932  
9992   13.3200  
9993   72.9480  

[9994 rows x 12 columns]

In [52]:
# by the above program mentioned, the selected column can be deleted but only temporary as you can see below, Postal Code is not removed from data.

In [51]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

In [53]:
df.drop('Postal Code',axis = 1,inplace = True)

In [54]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
0     South        Furniture    Bookcases  261.9600         2      0.00   
1     South        Furniture       Chairs  731.9400         3      0.00   
2      West  Office Supplies       Labels   14.6200         2      0.00   
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
...     ...              ...          ...       ...       ...       ...   
9989  South        Furniture  Furnishings   25.2480         3      0.20   
9990   West        Furniture  Furnishings   91.9600         2      0.00   
9991   West       Technology       Phones  258.5760         2      0.20   
9992   West  Office Supplies        Paper   29.6000         4      0.00   
9993   West  Office Supplies   Appliances  243.1600         2      0.00   

        Profit  
0      41.9136  
1     219.5820  
2       6.8714  
3    -383.0310  
4       2.5164  
...        ...  
9989    4.1028  
9990   15.6332  
9991   19.3932  
9992   13.3200  
9993   72.9480  

[9994 rows x 12 columns]

In [1]:
# now it is permenantly deleted, as inplace works to save the working directly into data. True means to remove whereas False means not to remove.

In [64]:
# here no axis required bcz rows index is 0 and bydefault it takes 0.

In [58]:
df.drop(9993,inplace = True)

In [59]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9988  Standard Class  Corporate  United States           Athens     Georgia   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
0     South        Furniture    Bookcases  261.9600         2      0.00   
1     South        Furniture       Chairs  731.9400         3      0.00   
2      West  Office Supplies       Labels   14.6200         2      0.00   
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
...     ...              ...          ...       ...       ...       ...   
9988  South       Technology       Phones  206.1000         5      0.00   
9989  South        Furniture  Furnishings   25.2480         3      0.20   
9990   West        Furniture  Furnishings   91.9600         2      0.00   
9991   West       Technology       Phones  258.5760         2      0.20   
9992   West  Office Supplies        Paper   29.6000         4      0.00   

        Profit  
0      41.9136  
1     219.5820  
2       6.8714  
3    -383.0310  
4       2.5164  
...        ...  
9988   55.6470  
9989    4.1028  
9990   15.6332  
9991   19.3932  
9992   13.3200  

[9993 rows x 12 columns]

In [60]:
df.drop(9990,inplace = True)

In [61]:
df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9987  Standard Class  Corporate  United States           Athens     Georgia   
9988  Standard Class  Corporate  United States           Athens     Georgia   
9989    Second Class   Consumer  United States            Miami     Florida   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
0     South        Furniture    Bookcases  261.9600         2      0.00   
1     South        Furniture       Chairs  731.9400         3      0.00   
2      West  Office Supplies       Labels   14.6200         2      0.00   
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
...     ...              ...          ...       ...       ...       ...   
9987  South       Technology  Accessories   79.9900         1      0.00   
9988  South       Technology       Phones  206.1000         5      0.00   
9989  South        Furniture  Furnishings   25.2480         3      0.20   
9991   West       Technology       Phones  258.5760         2      0.20   
9992   West  Office Supplies        Paper   29.6000         4      0.00   

        Profit  
0      41.9136  
1     219.5820  
2       6.8714  
3    -383.0310  
4       2.5164  
...        ...  
9987   28.7964  
9988   55.6470  
9989    4.1028  
9991   19.3932  
9992   13.3200  

[9992 rows x 12 columns]

In [63]:
# if you want to remove headers.

In [62]:
df2 = pd.read_csv('retail supermarket.csv',header=None)
df2

0          1              2                3           4   \
0          Ship Mode    Segment        Country             City       State   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class   Consumer  United States        Henderson    Kentucky   
3       Second Class  Corporate  United States      Los Angeles  California   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9990    Second Class   Consumer  United States            Miami     Florida   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993  Standard Class   Consumer  United States       Costa Mesa  California   
9994    Second Class   Consumer  United States      Westminster  California   

               5       6                7             8         9         10  \
0     Postal Code  Region         Category  Sub-Category     Sales  Quantity   
1           42420   South        Furniture     Bookcases    261.96         2   
2           42420   South        Furniture        Chairs    731.94         3   
3           90036    West  Office Supplies        Labels     14.62         2   
4           33311   South        Furniture        Tables  957.5775         5   
...           ...     ...              ...           ...       ...       ...   
9990        33180   South        Furniture   Furnishings    25.248         3   
9991        92627    West        Furniture   Furnishings     91.96         2   
9992        92627    West       Technology        Phones   258.576         2   
9993        92627    West  Office Supplies         Paper      29.6         4   
9994        92683    West  Office Supplies    Appliances    243.16         2   

            11        12  
0     Discount    Profit  
1            0   41.9136  
2            0   219.582  
3            0    6.8714  
4         0.45  -383.031  
...        ...       ...  
9990       0.2    4.1028  
9991         0   15.6332  
9992       0.2   19.3932  
9993         0     13.32  
9994         0    72.948  

[9995 rows x 13 columns]

In [65]:
# If you want your 1st row to be shown as header.

In [66]:
df3 = pd.read_csv('retail supermarket.csv',header=1)
df3

Second Class   Consumer  United States        Henderson    Kentucky  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class  Corporate  United States      Los Angeles  California   
2     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States      Los Angeles  California   
...              ...        ...            ...              ...         ...   
9988    Second Class   Consumer  United States            Miami     Florida   
9989  Standard Class   Consumer  United States       Costa Mesa  California   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992    Second Class   Consumer  United States      Westminster  California   

      42420  South        Furniture    Bookcases    261.96  2     0   41.9136  
0     42420  South        Furniture       Chairs  731.9400  3  0.00  219.5820  
1     90036   West  Office Supplies       Labels   14.6200  2  0.00    6.8714  
2     33311  South        Furniture       Tables  957.5775  5  0.45 -383.0310  
3     33311  South  Office Supplies      Storage   22.3680  2  0.20    2.5164  
4     90032   West        Furniture  Furnishings   48.8600  7  0.00   14.1694  
...     ...    ...              ...          ...       ... ..   ...       ...  
9988  33180  South        Furniture  Furnishings   25.2480  3  0.20    4.1028  
9989  92627   West        Furniture  Furnishings   91.9600  2  0.00   15.6332  
9990  92627   West       Technology       Phones  258.5760  2  0.20   19.3932  
9991  92627   West  Office Supplies        Paper   29.6000  4  0.00   13.3200  
9992  92683   West  Office Supplies   Appliances  243.1600  2  0.00   72.9480  

[9993 rows x 13 columns]

# Logical Operators in Pandas:
    * & (and)
    * | (or)
    * ~ (not)    tilde

In [67]:
# In Pandas show the sign irrespective of writing all as we did in python.

In [68]:
# SMID = Single Instruction and Multiple Data Explore.

In [69]:
# As your given instructions will just be single, so no if else statement is used in Pandas.

In [70]:
df['State'] == 'Florida'

0       False
1       False
2       False
3        True
4        True
        ...  
9987    False
9988    False
9989     True
9991    False
9992    False
Name: State, Length: 9992, dtype: bool

In [71]:
df[df['State'] == 'Florida']

Ship Mode      Segment        Country              City    State  \
3     Standard Class     Consumer  United States   Fort Lauderdale  Florida   
4     Standard Class     Consumer  United States   Fort Lauderdale  Florida   
43    Standard Class    Corporate  United States         Melbourne  Florida   
231   Standard Class  Home Office  United States             Tampa  Florida   
232   Standard Class  Home Office  United States             Tampa  Florida   
...              ...          ...            ...               ...      ...   
9818  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9819  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9820  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9821  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9989    Second Class     Consumer  United States             Miami  Florida   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
43    South  Office Supplies      Storage   95.6160         2      0.20   
231   South        Furniture       Tables  233.8600         2      0.45   
232   South        Furniture       Tables  620.6145         3      0.45   
...     ...              ...          ...       ...       ...       ...   
9818  South  Office Supplies          Art    4.4480         2      0.20   
9819  South  Office Supplies       Labels    9.0000         3      0.20   
9820  South  Office Supplies        Paper   42.2400        10      0.20   
9821  South  Office Supplies      Binders   18.2640         2      0.70   
9989  South        Furniture  Furnishings   25.2480         3      0.20   

        Profit  
3    -383.0310  
4       2.5164  
43      9.5616  
231  -102.0480  
232  -248.2458  
...        ...  
9818    0.3336  
9819    3.1500  
9820   13.2000  
9821  -13.3936  
9989    4.1028  

[383 rows x 12 columns]

In [72]:
df['Category'].unique()

array(['Furniture', 'Office Supplies', 'Technology'], dtype=object)

In [73]:
df[df['Category'] == 'Technology']

Ship Mode    Segment        Country           City       State  \
7     Standard Class   Consumer  United States    Los Angeles  California   
11    Standard Class   Consumer  United States    Los Angeles  California   
19      Second Class   Consumer  United States  San Francisco  California   
26      Second Class   Consumer  United States    Los Angeles  California   
35       First Class  Corporate  United States     Richardson       Texas   
...              ...        ...            ...            ...         ...   
9983  Standard Class   Consumer  United States   Grand Rapids    Michigan   
9986  Standard Class   Consumer  United States    Los Angeles  California   
9987  Standard Class  Corporate  United States         Athens     Georgia   
9988  Standard Class  Corporate  United States         Athens     Georgia   
9991  Standard Class   Consumer  United States     Costa Mesa  California   

       Region    Category Sub-Category     Sales  Quantity  Discount    Profit  
7        West  Technology       Phones   907.152         6       0.2   90.7152  
11       West  Technology       Phones   911.424         4       0.2   68.3568  
19       West  Technology       Phones   213.480         3       0.2   16.0110  
26       West  Technology  Accessories    90.570         3       0.0   11.7741  
35    Central  Technology       Phones  1097.544         7       0.2  123.4737  
...       ...         ...          ...       ...       ...       ...       ...  
9983  Central  Technology       Phones    97.980         2       0.0   27.4344  
9986     West  Technology  Accessories    36.240         1       0.0   15.2208  
9987    South  Technology  Accessories    79.990         1       0.0   28.7964  
9988    South  Technology       Phones   206.100         5       0.0   55.6470  
9991     West  Technology       Phones   258.576         2       0.2   19.3932  

[1847 rows x 12 columns]

In [74]:
df[(df['State'] == 'Florida') & (df['Category'] == 'Technology')]

Ship Mode      Segment        Country              City    State  \
235   Standard Class  Home Office  United States             Tampa  Florida   
314   Standard Class    Corporate  United States  Saint Petersburg  Florida   
385   Standard Class     Consumer  United States    Pembroke Pines  Florida   
435   Standard Class     Consumer  United States      Jacksonville  Florida   
633     Second Class     Consumer  United States             Miami  Florida   
...              ...          ...            ...               ...      ...   
8921  Standard Class    Corporate  United States      Coral Gables  Florida   
8957     First Class     Consumer  United States           Hialeah  Florida   
9138  Standard Class     Consumer  United States  Saint Petersburg  Florida   
9276    Second Class    Corporate  United States      Delray Beach  Florida   
9443  Standard Class  Home Office  United States      Jacksonville  Florida   

     Region    Category Sub-Category    Sales  Quantity  Discount   Profit  
235   South  Technology  Accessories  617.976         3       0.2  -7.7247  
314   South  Technology       Phones  178.384         2       0.2  22.2980  
385   South  Technology  Accessories   83.976         3       0.2  -1.0497  
435   South  Technology  Accessories  247.800         5       0.2 -18.5850  
633   South  Technology  Accessories   17.880         3       0.2   2.4585  
...     ...         ...          ...      ...       ...       ...      ...  
8921  South  Technology       Phones   52.680         3       0.2  19.7550  
8957  South  Technology     Machines  120.000         2       0.5  -7.2000  
9138  South  Technology       Phones  180.960         5       0.2  13.5720  
9276  South  Technology  Accessories   47.920         2       0.2  11.9800  
9443  South  Technology       Phones  219.184         2       0.2  19.1786  

[69 rows x 12 columns]

In [75]:
df[(df['State'] == 'Florida') | (df['State'] == 'California')]

Ship Mode      Segment        Country             City       State  \
2       Second Class    Corporate  United States      Los Angeles  California   
3     Standard Class     Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class     Consumer  United States  Fort Lauderdale     Florida   
5     Standard Class     Consumer  United States      Los Angeles  California   
6     Standard Class     Consumer  United States      Los Angeles  California   
...              ...          ...            ...              ...         ...   
9979  Standard Class  Home Office  United States      Los Angeles  California   
9986  Standard Class     Consumer  United States      Los Angeles  California   
9989    Second Class     Consumer  United States            Miami     Florida   
9991  Standard Class     Consumer  United States       Costa Mesa  California   
9992  Standard Class     Consumer  United States       Costa Mesa  California   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
2      West  Office Supplies       Labels   14.6200         2      0.00   
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
5      West        Furniture  Furnishings   48.8600         7      0.00   
6      West  Office Supplies          Art    7.2800         4      0.00   
...     ...              ...          ...       ...       ...       ...   
9979   West  Office Supplies      Binders  437.4720        14      0.20   
9986   West       Technology  Accessories   36.2400         1      0.00   
9989  South        Furniture  Furnishings   25.2480         3      0.20   
9991   West       Technology       Phones  258.5760         2      0.20   
9992   West  Office Supplies        Paper   29.6000         4      0.00   

        Profit  
2       6.8714  
3    -383.0310  
4       2.5164  
5      14.1694  
6       1.9656  
...        ...  
9979  153.1152  
9986   15.2208  
9989    4.1028  
9991   19.3932  
9992   13.3200  

[2382 rows x 12 columns]

In [76]:
(df['State'] == 'Florida') ~(df['Category'] == 'Technology')

SyntaxError: invalid syntax (3476291666.py, line 1)

In [77]:
(df['State'] == 'Florida') & ~(df['Category'] == 'Technology')

0       False
1       False
2       False
3        True
4        True
        ...  
9987    False
9988    False
9989     True
9991    False
9992    False
Length: 9992, dtype: bool

In [78]:
df[(df['State'] == 'Florida') & ~(df['Category'] == 'Technology')]

Ship Mode      Segment        Country              City    State  \
3     Standard Class     Consumer  United States   Fort Lauderdale  Florida   
4     Standard Class     Consumer  United States   Fort Lauderdale  Florida   
43    Standard Class    Corporate  United States         Melbourne  Florida   
231   Standard Class  Home Office  United States             Tampa  Florida   
232   Standard Class  Home Office  United States             Tampa  Florida   
...              ...          ...            ...               ...      ...   
9818  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9819  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9820  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9821  Standard Class    Corporate  United States  Saint Petersburg  Florida   
9989    Second Class     Consumer  United States             Miami  Florida   

     Region         Category Sub-Category     Sales  Quantity  Discount  \
3     South        Furniture       Tables  957.5775         5      0.45   
4     South  Office Supplies      Storage   22.3680         2      0.20   
43    South  Office Supplies      Storage   95.6160         2      0.20   
231   South        Furniture       Tables  233.8600         2      0.45   
232   South        Furniture       Tables  620.6145         3      0.45   
...     ...              ...          ...       ...       ...       ...   
9818  South  Office Supplies          Art    4.4480         2      0.20   
9819  South  Office Supplies       Labels    9.0000         3      0.20   
9820  South  Office Supplies        Paper   42.2400        10      0.20   
9821  South  Office Supplies      Binders   18.2640         2      0.70   
9989  South        Furniture  Furnishings   25.2480         3      0.20   

        Profit  
3    -383.0310  
4       2.5164  
43      9.5616  
231  -102.0480  
232  -248.2458  
...        ...  
9818    0.3336  
9819    3.1500  
9820   13.2000  
9821  -13.3936  
9989    4.1028  

[314 rows x 12 columns]

In [79]:
df[(df['State'] == 'Florida') | ~(df['Category'] == 'Technology')]

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9982  Standard Class   Consumer  United States     Grand Rapids    Michigan   
9984  Standard Class   Consumer  United States       Long Beach    New York   
9985  Standard Class   Consumer  United States       Long Beach    New York   
9989    Second Class   Consumer  United States            Miami     Florida   
9992  Standard Class   Consumer  United States       Costa Mesa  California   

       Region         Category Sub-Category     Sales  Quantity  Discount  \
0       South        Furniture    Bookcases  261.9600         2      0.00   
1       South        Furniture       Chairs  731.9400         3      0.00   
2        West  Office Supplies       Labels   14.6200         2      0.00   
3       South        Furniture       Tables  957.5775         5      0.45   
4       South  Office Supplies      Storage   22.3680         2      0.20   
...       ...              ...          ...       ...       ...       ...   
9982  Central  Office Supplies        Paper   35.5600         7      0.00   
9984     East  Office Supplies       Labels   31.5000        10      0.00   
9985     East  Office Supplies     Supplies   55.6000         4      0.00   
9989    South        Furniture  Furnishings   25.2480         3      0.20   
9992     West  Office Supplies        Paper   29.6000         4      0.00   

        Profit  
0      41.9136  
1     219.5820  
2       6.8714  
3    -383.0310  
4       2.5164  
...        ...  
9982   16.7132  
9984   15.1200  
9985   16.1240  
9989    4.1028  
9992   13.3200  

[8214 rows x 12 columns]

In [80]:
df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
                  index=['cobra', 'viper', 'sidewinder'],
                   columns=['max_speed', 'shield'])
df

max_speed  shield
cobra               1       2
viper               4       5
sidewinder          7       8